### ISSUE TITLE
Put plugins in own repos

### ISSUE THREAD

**Issue by morganney** [DataTables/Plugins#291]
Would you be able to put the plugins in their own repositories with a package.json file so that they can be installed with npm using a [github url](https://docs.npmjs.com/files/package.json#github-urls)? Right now only extensions seem manageable with npm.

**Comment by DataTables** [comment]
Its something I'm going to look at doing, but I need to write some scripts to do it as there will be hundreds of repos (which seems somewhat redundant to me for everything other than npm).

**Comment by Tol1** [comment]
What about multiple npm packages, one for every plugin, instead of multiple repositories. They can still sit on same repository. This needs package.json to root folder of every plugin, and (possibly a script to) publish every individual plugin folder as npm package.

**Comment by julmot** [comment]
@Tol1 Every plugin needs its own unit tests, documentation and version. If you save all plugins in one repository and would e.g. execute all tests of them and just one is failing, then this would block further releases for all plugins.

**Comment by Tol1** [comment]
Well, how that differs current situation? I was just proposing a way to improve current situation, where main functionality can be easily installed via npm, but plugins cannot.
Yes, i can use github url with version tag, but this way I lock version and must manually check if there's new compatible version, because github urls don't support semver notations.
I understand those difficulties on managing plugins, but thought it wouldn't differ on how they are managed now. Own repositories for each is of course better solution if you have time to build it that way

**Comment by julmot** [comment]
@Tol1 Yes, from looking at the current solution this would still be better. But, when changing the structure in general we should find the best possible solution imho, not trying to improve the current solution with the least effort.

**Comment by morganney** [comment]
Github urls support a commit-ish suffix to help lock down a specific snapshot of the plugin in your application. The script used to create the repositories can be extended slightly to publish each plugin to npm where semver and installs with yarn become possible.

**Comment by holtkamp** [comment]
@DataTables what would you consider to be the advised way to "install" plugins?

I would prefer to be able to actually "install" them and not rely on CDNs, so local development can continue while working offline...

**Comment by DataTables** [comment]
Three options spring to mind:

1. Copy the code from this repo and store it in your own
2. Clone this repo and pull in the files you need
3. Install the npm package `datatables.net-plugins` and pull in the files you need - e.g.

```
import 'datatables.net-plugins/api/average()';
```

Option three is probably the best, but it assumes you are using npm / yarn.

Allan

**Comment by holtkamp** [comment]
Thanks for the swift and clear response!

Somehow I missed this repository when listing all possible packages that can be installed using NPM / yarn, bower etc. 

FYI, I use Composer combined with the [Composer Asset Plugin](https://github.com/fxpio/composer-asset-plugin) which allows to install all plugins using `composer require bower-asset/datatables.net-plugins` which works nice. 

Off-topic: a notification message is displayed though:

```
  - Installing bower-asset/datatables.net-plugins (1.10.20): Loading from cache
    Failed to execute unzip -qq  '/project/assets/vendor/datatables.net-plugins/e3e63769c2ac5b9624f14f7013a31ca7' -d '/project/vendor/composer/4e661ac9'

replace /project/vendor/composer/4e661ac9/DataTables-Plugins-dfb3959/features/slidingChild/datatables.slidingChild.min.js? [y]es, [n]o, [A]ll, [N]one, [r]ename:  NULL
(EOF or read error, treating as "[N]one" ...)

    The archive may contain identical file names with different capitalization (which fails on case insensitive filesystems)
    Unzip with unzip command failed, falling back to ZipArchive class
```
 So there might be some capitalization issues in the repository?

PS: @DataTables did you consider to signup for the [GitHub Sponsor program](https://github.com/sponsors)? Christmas is coming 👼

**Comment by DataTables** [comment]
Thanks! Fixed in https://github.com/DataTables/Plugins/commit/90ae50b661204a1cd5f751b1717b120254e8572e.

And thanks for the link to the sponsor program. We've not signed up to that yet. I need to think about it a bit more - while the majority of our work is open source, we do also have our Editor component which is commercial software and that's where our revenue currently comes from. So the best way to support DataTables is to purchase Editor :-).

**Comment by holtkamp** [comment]
>And thanks for the link to the sponsor program. We've not signed up to that yet. I need to think about it a bit more - while the majority of our work is open source, we do also have our Editor component which is commercial software and that's where our revenue currently comes from. So the best way to support DataTables is to purchase Editor :-).

Sure, point is: some people might not need the Editor and still want to express their gratitude for all the nice functionality, documentation and support with an amount of money they could afford...
